executor/src/witgen/jit/affine_symbolic_expression.rs

```diff
@@ -0,0 +1,532 @@
+        AffineSymbolicExpression {
+            coefficients: Default::default(),
+            offset: k,
+        }
+    }
+}
+
+impl<T: FieldElement, V: Ord + Clone + Display> AffineSymbolicExpression<T, V> {
+    pub fn from_known_variable(var: V) -> Self {
+        SymbolicExpression::from_var(var).into()
+    }
+    pub fn from_unknown_variable(var: V) -> Self {
+        AffineSymbolicExpression {
+            coefficients: [(var, T::from(1).into())].into_iter().collect(),
+            offset: SymbolicExpression::from(T::from(0)),
+        }
+    }
+    pub fn from_number(n: T) -> Self {
+        SymbolicExpression::from(n).into()
+    }
+
+    /// If this expression does not contain unknown variables, returns the symbolic expression.
+    pub fn try_to_known(&self) -> Option<&SymbolicExpression<T, V>> {
+        if self.coefficients.is_empty() {
+            Some(&self.offset)
+        } else {
+            None
+        }
+    }
+
+    /// Tries to multiply this expression with another one.
+    /// Returns `None` if the result would be quadratic.
+    pub fn try_mul(&self, other: &Self) -> Option<Self> {
+        if !self.coefficients.is_empty() && !other.coefficients.is_empty() {
+            return None;
+        }
+        let (multiplier, coefficients, offset) = if self.coefficients.is_empty() {
+            (&self.offset, &other.coefficients, &other.offset)
+        } else {
+            (&other.offset, &self.coefficients, &self.offset)
+        };
+        let coefficients = coefficients
+            .iter()
+            .map(|(var, coeff)| (var.clone(), coeff * multiplier))
+            .collect();
+        let offset = offset * multiplier;
+        Some(AffineSymbolicExpression {
+            coefficients,
+            offset,
+        })
+    }
+
+    /// Solves the equation `self = 0` and returns how to compute the solution.
+    pub fn solve(
+        &self,
+        range_constraints: &impl RangeConstraintSet<V, T>,
+    ) -> Result<Vec<Effect<T, V>>, EvalError<T>> {
+        Ok(match self.coefficients.len() {
+            0 => {
+                if self.offset.is_known_zero() {
+                    vec![]
+                } else {
+                    return Err(EvalError::ConstraintUnsatisfiable(self.to_string()));
+                }
+            }
+            1 => {
+                let (var, coeff) = self.coefficients.iter().next().unwrap();
+                assert!(
+                    !coeff.is_known_zero(),
+                    "Zero coefficient has not been removed."
+                );
+                if coeff.is_known_nonzero() {
+                    let value = self.offset.field_div(&-coeff);
+                    vec![Effect::Assignment(var.clone(), value)]
+                } else {
+                    // We can only solve this if we know that the coefficient cannot be zero.
+                    // TODO We could do this by adding a runtime condition
+                    vec![]
```

Comment:
Wait, so now (with the change of cad7f37b3b3e02686aab364bc7482b89310bf707), it wouldn't be able to solve `A * B = 5` anymore for a known `B`, right? Because `B` would not have a compile-time range constraint, so `coeff.is_known_nonzero()` would be false?

I think the way it was before is fine, just that the `Assertion::assert_is_nonzero(coeff.clone())` was overly strict (which I think we can fix later).

Reply:
Hmm, I think it is fine like this for now, but it doesn't cover all cases. I think [this example](https://github.com/powdr-labs/powdr/pull/2212#discussion_r1879801242) still wouldn't work, because neither the numerator nor denominator would be known to be (non)zero at compile time.

For context, the affine expression would be:
`(-instr_add - instr_sub) * c + (instr_add * (a + b) + instr_sub * (a - b)) = 0`

Assuming binary constraints on the instruction flags, what we have here is that:
- `coeff == 0` $\Rightarrow$ `instr_add == 0 && instr_sub == 0` $\Rightarrow$ `offset == 0`.
- But we'll only know the instruction flags at runtime, so at runtime we need to check the value of `coeff` and:
  - Do nothing if it zero
  - Otherwise assign `c := -offset / coeff`

But let's iterate once we get there...

Reply:
Yep!